pkg/streamingpromql/operator/ring_buffer.go

```diff
@@ -26,13 +26,16 @@ func (b *RingBuffer) DiscardPointsBefore(t int64) {
 	}
 }
 
-// Points returns slices of the points in this buffer.
+// HeadAndTail returns slices of the points in this buffer.
 // Either or both slice could be empty.
-// Callers must not modify the values in the returned slices.
+// Callers must not modify the values in the returned slices or return them to a pool.
+// Calling HeadAndTail is more efficient than calling CopyPoints, as CopyPoints will create a new slice and copy all
+// points into the slice, whereas HeadAndTail returns a view into the internal state of this buffer.
+// The returned slices are no longer valid if this buffer is modified (eg. a point is added, or the buffer is reset).
 //
 // FIXME: the fact we have to expose this is a bit gross, but the overhead of calling a function with ForEach is terrible.
 // Perhaps we can use range-over function iterators (https://go.dev/wiki/RangefuncExperiment) once this is not experimental?
-func (b *RingBuffer) Points() ([]promql.FPoint, []promql.FPoint) {
+func (b *RingBuffer) HeadAndTail() ([]promql.FPoint, []promql.FPoint) {
```

Comment:
[non blocking] I'm wondering if you considered renaming `HeadAndTail()` to `UnsafePoints()`. They this function and `CopyPoints()` return the same data, but `HeadAndTail()` is the unsafe version and since it exposes the underlying buffers it's made of 2 buffers instead of 1. Don't have to follow my suggestion. I know many people doesn't like name methods "unsafe". I personally think it makes very explicit that you're using a potentially dangerous function.

Reply:
Good idea, changed in a405e08